pallene/checker.lua

```diff
@@ -451,9 +452,42 @@ function Checker:check_stat(stat)
             self:add_local(stat.decl)
             self:check_stat(stat.block)
         end)
+    elseif tag == "ast.Stat.ForIn" then
+        local rhs = stat.exps
+        self:expand_function_returns(3, rhs)
+
+        local expected_ret_types = {}
+        for _ = 1, #stat.decls do
+            table.insert(expected_ret_types, types.T.Any())
+        end
+
+        local iteratorfn = self:check_exp_verify(rhs[1], types.T.Function({
+            types.T.Any(), types.T.Any()
+        }, expected_ret_types), "loop iterator")
+
+        if not rhs[2] then
+            type_error(stat.loc, "Missing state argument in for-in loop")
+        else
+            rhs[2] = self:check_exp_verify(rhs[2], types.T.Any(), "state variable")
+        end
+
+        rhs[3] = rhs[3] and self:check_exp_verify(rhs[3], types.T.Any(), "control variable")
+
+        if #stat.decls ~= #iteratorfn._type.ret_types then
+            type_error(stat.decls[1].loc, "expected %d values, but function returns %d",
+                       #stat.decls, #iteratorfn._type.ret_types)
+        end
```

Comment:
One thing to watch out here if we set `expected_ret_types` to an array of `any` and then call `check_exp_verify`, then it will insert a type coercion and the return type of `iteratorfn` will literally be an array of `any` as well. Maybe that is a sign that we shouldn't be using `check_exp_verify` up there.

Reply:
Ah yes, so what do you suggest I do? Inline some code there that does the type checking or use another function?